DelegatingCompletableFuture_WrappingInvocationFutureTest

```
java.lang.AssertionError: expected:<java.lang.Object@7a5b4ab2> but was:<com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode@3f59cb9d>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.hazelcast.spi.impl.CompletableFutureAbstractTest.exceptionally_whenAsync(CompletableFutureAbstractTest.java:812)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/2389/testReport/junit/com.hazelcast.spi.impl/DelegatingCompletableFuture_WrappingInvocationFutureTest/exceptionally_whenAsync/

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java
@@ -486,6 +486,21 @@ public abstract class AbstractInvocationFuture<V> extends InternalCompletableFut
     }
 
     @Override
+    public InternalCompletableFuture<V> exceptionally(@Nonnull Function<Throwable, ? extends V> fn) {
+        requireNonNull(fn);
+        final InternalCompletableFuture<V> future = newCompletableFuture();
+        if (isDone()) {
+            unblockExceptionally(fn, future);
+        } else {
+            Object result = registerWaiter(new ExceptionallyNode<>(future, fn), null);
+            if (result != UNRESOLVED) {
+                unblockExceptionally(fn, future);
+            }
+        }
+        return future;
+    }
+
+    @Override
     public InternalCompletableFuture<V> toCompletableFuture() {
         return this;
     }
@@ -864,33 +879,19 @@ public abstract class AbstractInvocationFuture<V> extends InternalCompletableFut
         }
     }
 
-    @Override
-    public InternalCompletableFuture<V> exceptionally(@Nonnull Function<Throwable, ? extends V> fn) {
-        requireNonNull(fn);
+    private void unblockExceptionally(@Nonnull Function<Throwable, ? extends V> fn,
+                                      InternalCompletableFuture<V> future) {
         Object result = resolve(state);
-        final InternalCompletableFuture<V> future = newCompletableFuture();
-        for (; ; ) {
-            if (result != UNRESOLVED && isDone()) {
-                if (result instanceof ExceptionalResult) {
-                    Throwable throwable = ((ExceptionalResult) result).cause;
-                    try {
-                        V value = fn.apply(throwable);
-                        future.complete(value);
-                    } catch (Throwable t) {
-                        future.completeExceptionally(t);
-                    }
-                } else {
-                    future.complete((V) result);
-                }
-                return future;
-            } else {
-                result = registerWaiter(new ExceptionallyNode<>(future, fn), null);
-                if (result == UNRESOLVED) {
-                    return future;
-                } else {
-                    result = resolve(state);
-                }
+        if (result instanceof ExceptionalResult) {
+            Throwable throwable = ((ExceptionalResult) result).cause;
+            try {
+                V value = fn.apply(throwable);
+                future.complete(value);
+            } catch (Throwable t) {
+                future.completeExceptionally(t);
             }
+        } else {
+            future.complete((V) result);
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/DelegatingCompletableFuture.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/DelegatingCompletableFuture.java
@@ -408,7 +408,7 @@ public class DelegatingCompletableFuture<V> extends InternalCompletableFuture<V>
 
     @Override
     public CompletableFuture<V> exceptionally(Function<Throwable, ? extends V> fn) {
-        return  new DelegatingCompletableFuture<>(serializationService,
+        return new DelegatingCompletableFuture<>(serializationService,
                 future.exceptionally(fn));
     }
 

